verl/workers/rollout/schemas.py

```diff
@@ -82,127 +88,121 @@ class AsyncRolloutRequest(BaseModel):
     prompt_loss_mask: List[int]
     response_loss_mask: List[int]
     reward_scores: Dict[str, float]
+    max_prompt_len: int
     max_response_len: int = 8192
     max_model_len: int = 32768
 
-    format_config: dict = {
-        "chatml": {
-            "assistant_prefix_msg": "\n<|im_start|>assistant\n",
-            "assistant_suffix_msg": "<|im_end|>",
-            "tool_prefix_msg": "\n<|im_start|>tool\n",
-            "tool_suffix_msg": "<|im_end|>",
-        },
-        "qwen": {
-            "assistant_prefix_msg": "\n<|im_start|>assistant\n",
-            "assistant_suffix_msg": "<|im_end|>",
-            "merge_tool_response": True,
-            "tool_prefix_msg": "\n<|im_start|>user",
-            "tool_suffix_msg": "<|im_end|>",
-            "tool_response_prefix_msg": "\n<tool_response>\n",
-            "tool_response_suffix_msg": "\n</tool_response>",
-        }
-    }
-
-    def get_generation_prompt(self, tokenizer: PreTrainedTokenizer) -> list[int]:
-        return tokenizer.apply_chat_template(  # type: ignore
-            conversation=[msg.model_dump() for msg in self.messages],
-            tools=[tool.model_dump() for tool in self.tools] if self.tools else None,
-            add_generation_prompt=True,
-            tokenize=True,
-        )
+    tokenization_mode: Literal["fast", "full", "sanity_check"]
+    generation_prompt_ids: List[int]
+
+    @model_validator(mode="before")
+    @classmethod
+    def initialize_request(cls, values):
+        if not (messages := values.get("messages")):
+            raise ValueError("messages is required for AsyncRolloutRequest initialization")
+        if not (max_prompt_len := values.get("max_prompt_len")):
+            raise ValueError("max_prompt_len is required for AsyncRolloutRequest initialization")
+        if not (tokenizer := values.pop("tokenizer", None)):
+            raise ValueError("tokenizer is required for AsyncRolloutRequest initialization")
+
+        values["messages"] = [Message.model_validate(msg) for msg in messages]
+        values["messages_dumps"] = [msg.model_dump() for msg in values["messages"]]
+
+        if tool_schemas := values.get("tool_schemas"):
+            tools = values["tools"] = [tool.model_dump() for tool in tool_schemas]
```

Comment:
Can tools []/None be used directly to switch here? Rather than use if-else. And I'm a little confused why we need pre configured input_ids and attention mask here, it seems can be supported as lazy tokenization (only do tokenize when missing) ?